src/bucket/Bucket.cpp

```diff
@@ -181,22 +181,79 @@ Bucket::getBucketEntry(LedgerKey const& k)
     return std::nullopt;
 }
 
+void
+Bucket::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+                 std::vector<LedgerEntry>& result)
+{
+    std::unordered_map<LedgerKey, std::unordered_set<Hash>> lkToTx;
+    std::unordered_map<Hash, uint32_t> txReadBytes;
+    loadKeys(keys, result, lkToTx, txReadBytes);
+}
+
 // When searching for an entry, BucketList calls this function on every bucket.
 // Since the input is sorted, we do a binary search for the first key in keys.
 // If we find the entry, we remove the found key from keys so that later buckets
 // do not load shadowed entries. If we don't find the entry, we do not remove it
 // from keys so that it will be searched for again at a lower level.
 void
-Bucket::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
-                 std::vector<LedgerEntry>& result)
+Bucket::loadKeys(
+    std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+    std::vector<LedgerEntry>& result,
+    std::unordered_map<LedgerKey, std::unordered_set<Hash>>& lkToTx,
+    std::unordered_map<Hash, uint32_t>& txReadBytes)
 {
     ZoneScoped;
+    auto hasTxnWithReadQuota = [&](auto key) {
+        if (lkToTx.empty() || txReadBytes.empty() ||
+            /* classic */ lkToTx.count(key) == 0)
+        {
+            return true;
+        }
+        // Ensure at least one transaction has read quota for this key.
+        for (auto txn : lkToTx[key])
+        {
+            // TODO maybe if (txReadBytes[txn] >=  getIndex().getPageSize())
+            // Ensure txReadBytes[txn] is non-zero
+            if (txReadBytes[txn] > 0)
+            {
+                return true;
+            }
+        }
+        return false;
+    };
+    auto updateReadQuotaForLK = [&](auto key, auto entry) {
+        if (lkToTx.empty() || txReadBytes.empty() ||
+            /* classic */ lkToTx.count(key) == 0)
+        {
+            return;
+        }
+        // Update read quota for all txns waiting on this key.
+        auto size = xdr::xdr_size(entry);
+        for (auto txn : lkToTx[key])
+        {
+            // Prevent underflow.
+            if (txReadBytes[txn] < size)
+            {
+                txReadBytes[txn] = 0;
+            }
+            else
+            {
+                txReadBytes[txn] -= size;
+            }
+        }
+    };
 
     auto currKeyIt = keys.begin();
     auto const& index = getIndex();
     auto indexIter = index.begin();
     while (currKeyIt != keys.end() && indexIter != index.end())
     {
+        if (!hasTxnWithReadQuota(*currKeyIt))
+        {
+            // If no txn has read quota for this key, skip it.
+            currKeyIt = keys.erase(currKeyIt);
```

Comment:
There's a subtle bug here. The way that Bucket loads work, it is assumed that every key is loaded. This means that if there is a key in `keys`, and a `LedgerEntry` for that key is not inserted into `result`, it is assumed that the key was loaded, but does not exist. In turn, the `LedgerTxn` that called `loadKeys` assumes that if it doesn't see a result, the result is nullptr. This is a very significant issue, as the cache will have incorrect state. If you run out of resources before you can load `k`, but `k` exists, the cache will have the mapping `k -> nullptr`. If during TX apply another operation loads this key (say a classic OP), the cache will hit, and the operation will execute as if the entry does not exist when in reality it does.

I think you will need to add some way to communicate back to the `LedgerTxn` that the key is not null, but rather that it was not loaded. You could accomplish this by passing in `std::vector<LedgerKey>& notLoaded` to prefetch. After prefetch returns, if a key is in `notLoaded`, that key will not be loaded in the cache. While sublte, not being loaded into the cache is very different than being loaded into the cache as null.

Reply:
Hmm yeah good catch. I think explicitly storing the excluded keys in notLoaded is a good approach.